### PR TITLE
ImageStore statement in IR

### PR DESCRIPTION
**Pull request by kvark**

### CODE REVIEW

**Review by Napokue — APPROVED** [review]
LGTM, thank you! 

What did change that you removed the `crate::` or `spirv::` in some files? I believe we once talked about this and you said that you have a preference over using the full qualifier instead of shortening it.

**Comment by Diggsey** [comment]
I tested using `ImageStore` on DX12 and got the following error:
> ERROR gfx_backend_dx12::device > D3DCompile error 80004005: C:\Users\diggs\Projects\lbfs\Shader@0x000002148BA21E10(14,5-53): error X3025: l-value specifies const object

I'm unsure if this is an issue with naga or with spirv-cross though (currently attempting to test in vulkan to see if the spv output works when used directly)

**Comment by Diggsey** [comment]
The `textureStore` issue appears to be a problem with `spirv-cross`, as I can use the .spv with vulkan directly and it works.

However, I found an issue with `textureLoad`, where naga appears to be generating invalid SPIRV. It seems any usage of this function generates an .spv file which I cannot disassemble with `spriv-dis` due to the following error:

> error: 101: Id 14 is defined more than once

If I try to *use* the shader, I get a segfault.

**Comment by kvark** [comment]
Thanks @Diggsey ! I'll look into this and make sure the generated SPIR-V is correct. If you have a shader in mind that would use select/switch/ imageStore, I'd be happy to add a snapshot test for it.

**Comment by Diggsey** [comment]
@kvark Here's the code I was testing: it's basically the bare minimum texture load/store example, copying one image to another.

```wgsl

// CS inputs
[[builtin(global_invocation_id)]] var<in> global_id: vec3<u32>;

// CS bindings
[[group(0), binding(0)]]
var bound_src: [[access(read)]] texture_storage_2d<rgba32float>;
[[group(0), binding(1)]]
var bound_dst: [[access(write)]] texture_storage_2d<rgba32float>;

[[stage(compute), workgroup_size(1)]]
fn compute() {
    var coord: vec2<i32> = vec2<i32>(global_id.x, global_id.y);
    var value: vec4<f32> = textureLoad(bound_src, coord);
    textureStore(bound_dst, coord, value);
}
```

**Comment by kvark** [comment]
@Diggsey thank you!
I was looking for more of a real-world example, actually. But I'll be happy to test your shader code locally, at least.

@Napokue thank you for the review!
> What did change that you removed the crate:: or spirv:: in some files? I believe we once talked about this and you said that you have a preference over using the full qualifier instead of shortening it.

It's the new warning I enabled about trivial casts. In the places I fixed the thing behind `crate::xxx` was already imported into the scope, so `crate::` was unnecessary.
Our code is not fully consistent on the code style, so I wrote https://github.com/gfx-rs/naga/wiki/Code-style to help.

File: src/back/spv/writer.rs [hunk 1]
```diff
@@ -1510,6 +1594,44 @@ impl Writer {
...
+                    ir_function,
+                    coordinate,
+                    array_index,
+                    block,
+                    function,
+                )?;
+
+                let id = self.generate_id();
+                let mut instruction =
+                    Instruction::image_read(id, result_type_id, image_id, coordinate_id);
```
Comment (Diggsey) [inline src/back/spv/writer.rs:1616]: It looks like `id` and `result_type_id` are transposed here, and that's causing the invalid output.

Comment (kvark) [reply src/back/spv/writer.rs]: this is very helpful, thank you!

**Comment by Diggsey** [comment]
When I made that change locally, my example ran successfully, so toy example though it may be, I think that's all that's required to implement cool things like Lattice-Boltzmann fluid simulation :)

**Comment by kvark** [comment]
wow nice!
there are still fixes needed though, I'm seeing:
> error: line 40: Target of NonWritable decoration is invalid: must point to a storage image, uniform block, or storage buffer
  %bound_src = OpVariable %_ptr_UniformConstant_29 UniformConstant

**Comment by Diggsey** [comment]
Weird, I'm running this exact code and it's working fine? 😕

**Comment by kvark** [comment]
Ok, it's all fixed now
Note that your example should use 
> var coord: vec2<i32> = vec2<i32>(vec2<u32>(global_id.x, global_id.y));

**Comment by kvark** [comment]
spirv-cross just ignores some of the flags, it's not validating, hence it seems to work for you.
anyway, I fixed everything. Thank you for quick testing! It's really making the development much easier ❤️

**Pull request merged**